src/pudl/analysis/mcoe.py

```diff
@@ -475,23 +458,16 @@ def mcoe(
     )
 
     # Combine MCOE derived values with all the generator attributes:
-    mcoe_out = (
-        pd.merge(
-            left=(
-                pudl_out.gens_eia860()
-                .assign(year=lambda x: x.report_date.dt.year)
-                .drop("report_date", axis="columns")
-            ),
-            right=mcoe_out.assign(year=lambda x: x.report_date.dt.year),
-            # This "how" determines whether MCOE or gens_eia860 is the backbone
-            how="left" if all_gens else "right",
-            on=["year", "plant_id_eia", "generator_id"],
-        )
-        .astype({"year": str})
-        .assign(report_date=lambda x: x.report_date.fillna(pd.to_datetime(x.year)))
-        .drop("year", axis="columns")
-        .pipe(pudl.validate.no_null_rows, df_name="mcoe_all_gens", thresh=0.9)
-    )
+    mcoe_out = pudl.helpers.full_timeseries_date_merge(
+        left=pudl_out.gens_eia860(),
+        right=mcoe_out,
+        on=["plant_id_eia", "generator_id"],
+        date_on=["year"],
+        how="outer" if all_gens else "right",
+        start=pudl_out.start_date,
+        end=pudl_out.end_date,
+        freq=pudl_out.freq,
+    ).pipe(pudl.validate.no_null_rows, df_name="mcoe_all_gens", thresh=0.9)
```

Comment:
@cmgosnell I changed this based on your [comment here](https://github.com/catalyst-cooperative/pudl/pull/1550#issuecomment-1106597818). Let me know if the parameters feel right here. I changed this so if `all_gens` is True an `outer` merge is performed instead of a `left` merge (using the keys from `gens_eia860`. I think they do the same thing since I was interpreting it as the generators in `mcoe_out` are a subset of the generators in `gens_eia860()`, but it intuitively made more sense to me to do an `outer` merge so that all the generators are included.